hpet: make it support EHL/PSE CPU

kwd-doodling opened the pull request:
- Change to 64-bit mode for better usages.
- Add Kconfig option to change min_delay and clock period.
- Add Kconfig option to set interrupt as level triggered for ARM CPUs
with NVIC, whose DTS interrupts setting has no "sense" cell.
- Make the change of HPET comparator register as a weak function, which
could be overwritten by EHL/PSE CPU to implement its special operations.

Signed-off-by: Dong Wang <dong.d.wang@intel.com>

dcpleung commented on drivers/timer/hpet.c:
```diff
@@ -83,29 +107,31 @@ static void hpet_isr(const void *arg)
 	val |= TCONF_INT_LEVEL;
 #endif
 
+	val = (val & ~(TCONF_PERIODIC | TCONF_FSB_EN)) | TCONF_INT_ENABLE;
```
For completeness, please also mask the 32-bit counter bit.

chen-png reviewed (approved):
looks good to me,  I checked the HPET timer code EHL PSE used before, it's ok with this change.

jhedberg commented:
@kwd-doodling could you fix the merge conflicts? They were introduced since the PR from @dcpleung was merged.

kwd-doodling committed:
timer: hpet: enable 64 bit mode for better usages

Get longer maximum timeout
Make HPET counter usable as timestamp

Signed-off-by: Dong Wang <dong.d.wang@intel.com>

kwd-doodling committed:
timer: hpet: make it support EHL/PSE CPU

Add config macro to set interrupt as level triggered for ARM CPUs
Merge all timer configures into one place, then no need to overwrite
hpet_timer_conf_get/set() functions in SoC layer
Make hpet_timer_comparator_set() as the only register access function
to implemented in the SoC layer

Signed-off-by: Dong Wang <dong.d.wang@intel.com>

dcpleung reviewed (changes requested):
You are removing the ability of the SoC layer to define its own functions (except one function). The ability is there for a reason. Please do not remove them.

kwd-doodling commented:
I've enabled this driver on both ISH and PSE and made it pass all timer tests.
until now only the comparator_set function is MUST to overwrite.
Timer configuration is WA by configuring only once.
if not doing so, I need to copy all of them into SoC layer.
But I think it's good to keep them still as functions to move into SoC layer when needed.

tbursztyka commented on drivers/timer/hpet.c:
```diff
@@ -208,9 +218,14 @@ static inline void hpet_timer_conf_set(uint32_t val)
+static inline void hpet_timer_comparator_set(uint64_t val)
 {
-	sys_write32(val, TIMER0_COMPARATOR_REG);
+#if CONFIG_X86_64
```
using if (IS_ENABLED(CONFIG_X86_64)) ... would be nice here.

kwd-doodling replied:
only X86_64 has sys_write64() function. #if MUST bs used, or there will be compiling error.

tbursztyka replied:
indeed

tbursztyka commented on drivers/timer/hpet.c:
```diff
@@ -224,30 +239,26 @@ static inline void hpet_timer_comparator_set(uint32_t val)
-	((MAX_TICKS - cyc_per_tick) / cyc_per_tick)
 #endif /* CONFIG_TIMER_READS_ITS_FREQUENCY_AT_RUNTIME */
 
+#define max_ticks ((int32_t)0x7fffffff)
```
Since you are removing the max_ticks variable above, just make the macro in capital letters at it should be.

kwd-doodling replied:
done.

tbursztyka commented on drivers/timer/hpet.c:
```diff
@@ -1,5 +1,5 @@
 /*
```
change you commit title to "drivers: timer: Enable 64 bit mode in HPET for better usage" or alike

kwd-doodling replied:
done.

tbursztyka commented on drivers/timer/hpet.c:
```diff
@@ -345,11 +352,16 @@ int sys_clock_driver_init(const struct device *dev)
 	reg &= ~TIMER_CONF_PERIODIC;
 	reg &= ~TIMER_CONF_FSB_EN;
-	reg |= TIMER_CONF_MODE32;
+	reg &= TIMER_CONF_MODE32;
```
so you are not setting this config mode then?
Or did you mean to have : reg &= ~TIMER_CONF_MODE32;  ?

kwd-doodling replied:
It's a mistake. fixed.

kwd-doodling commented:
@tbursztyka @dcpleung can you review again? Thanks.

tbursztyka commented on drivers/timer/hpet.c:
```diff
@@ -260,14 +256,20 @@ static void hpet_isr(const void *arg)
+	hpet_int_sts_set(TIMER0_INT_STS);
+#endif
+#else
+#if HPET_INT_LEVEL_TRIGGER
```
maybe all this #if/#else could be gathered into one macro function, that you could use like: 
```
#if NAME_OF_THE_MACRO()
        do_something();
#endif
```

You could reuse that below for level trigger as well. could help for readability.

kwd-doodling replied:
The thing is a little complicated, no easy way to gather into one macro.
In the new version I simplified it into one macro by changes from line 224 to 237. I also explained the background there.
Please review, thanks.

dcpleung reviewed (changes requested):
My previous comment still stands.

kwd-doodling commented:
> My previous comment still stands.
@dcpleung experiments on ISH and PSE show that only one of the functions is necessary. if keeping your code, I need to exactly copy all other functions into soc.h. It seems not good.

dcpleung commented:
On which ISH version have you tried?

kwd-doodling commented:
tried on ISH Chrome code base by switching zephyr to v2.6-branch. I've sent you an email to describe the steps.

dcpleung commented on drivers/timer/hpet.c:
```diff
@@ -378,15 +390,15 @@ void sys_clock_set_timeout(int32_t ticks, bool idle)
 
-	ticks = ticks == K_TICKS_FOREVER ? max_ticks : ticks;
-	ticks = CLAMP(ticks - 1, 0, (int32_t)max_ticks);
+	ticks = ticks == K_TICKS_FOREVER ? HPET_MAX_TICKS : ticks;
```
`MAX_TICKS` and `max_ticks` are both calculated value based on cycles per tick. And you are simply replacing them with `0x7FFFFFFF`. These are used so that we get at least one timer interrupt before the clock counter overflows. Simply replacing them with a constant is not right.

kwd-doodling replied:
before HPET works in 32-bit mode, so its MAX_TICK must be calculated at runtime.
now it's switched to 64-bit mode, 0x7fffffff is for sure supported.
